### PR TITLE
Make CLI more robust to discovery latency.

### PR DESCRIPTION
**Pull request by hidmic**
Split from #421. This pull request builds on top of #493 and ensures the daemon gets used throughout all CLI tools. It also updates some tests which showed to be flakey nonetheless (with and without this patch).

### CODE REVIEW

**Comment by hidmic** [comment]
With https://github.com/ros2/ros2cli/pull/494/commits/fac0a491aa2bd10c85ed295141f7d39fe20bb4bd, https://github.com/ros2/ros2cli/issues/488 should be now addressed.

**Comment by hidmic** [comment]
CI up to `ros2topic`, `ros2service`, `ros2action`, `ros2node` and `ros2component`:

- Linux [Build Status](https://ci.ros2.org/job/ci_linux/10531/)
- Linux-aarch64 [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/5989/)
- OSX [Build Status](https://ci.ros2.org/job/ci_osx/8578/)
- Windows [Build Status](https://ci.ros2.org/job/ci_windows/10398/)

File: ros2component/ros2component/api/__init__.py [hunk 1]
```diff
@@ -82,7 +81,10 @@ def get_container_components_info(*, node, remote_container_node_name):
         '{}/_container/list_nodes'.format(remote_container_node_name)
     )
     try:
-        list_nodes_client.wait_for_service()
+        if not list_nodes_client.wait_for_service(timeout_sec=5.0):
```
Comment (dirk-thomas) [inline ros2component/ros2component/api/__init__.py:84]: When used in a loop (as in `component list`) a blocking timeout of 5s can add up to a very long time for the user call.

File: ros2topic/ros2topic/verb/echo.py [hunk 2]
```diff
@@ -95,16 +97,20 @@ def main(args):
         callback = subscriber_cb_csv(truncate_length, args.no_arr, args.no_str)
     qos_profile = qos_profile_from_short_keys(
         args.qos_profile, reliability=args.qos_reliability, durability=args.qos_durability)
+    if args.message_type is None:
+        with NodeStrategy(args) as node:
```
Comment (dirk-thomas) [inline ros2topic/ros2topic/verb/echo.py:101]: This potentially creates a second direct node (beside the one created a few lines below) which doesn't seem like a good idea.

Same pattern below in multiple cases (even though they preexisted before).

**Comment by dirk-thomas** [comment]
> With fac0a49, #488 should be now addressed.

Please use a format which will ensure the referenced ticket is automatically closed when the PR is merged.

File: ros2action/test/test_cli.py [hunk 3]
```diff
@@ -251,8 +252,6 @@ def test_send_fibonacci_goal_with_feedback(self):
             assert action_command.wait_for_shutdown(timeout=10)
         assert action_command.exit_code == launch_testing.asserts.EXIT_OK
         assert launch_testing.tools.expect_output(
-            expected_lines=get_fibonacci_send_goal_output(
-                order=5, with_feedback=True
-            ),
-            text=action_command.output, strict=True
+            expected_lines=get_fibonacci_send_goal_output(order=5, with_feedback=1),
+            text=action_command.output, strict=False
```
Comment (dirk-thomas) [inline ros2action/test/test_cli.py:256]: Why not `strict` anymore?

Comment (hidmic) [reply ros2action/test/test_cli.py:256]: TL; DR because we cannot guarantee all action feedback will be received.

This bit is covered by https://github.com/ros2/ros2cli/pull/494/commits/b82bcd697c9bc5359694a558dcbd4475051d6e41. Test was flaky because it expected all published goal feedback to be printed. But the `ros2 action send_goal` exits once the goal result arrives, but actions' design does not (it cannot) guarantee all feedback will have arrived by the time the result response arrives. So I had to change the test to only expect the first piece of expected feedback and then look for the result -- which still does not provide any guarantees but it's much less likely to fail.

Having said that, we could try to match as many feedback messages as possible. But `expect_output()` can't do that now.

File: ros2component/ros2component/api/__init__.py (hunk 1 above)
Comment (hidmic) [reply ros2component/ros2component/api/__init__.py:84]: Fair enough. Timeout is a somewhat arbitrary, larger than zero quantity. What about `2.0` seconds?

File: ros2topic/ros2topic/verb/echo.py (hunk 2 above)
Comment (hidmic) [reply ros2topic/ros2topic/verb/echo.py:101]: That is true, you might end up creating node, destroying it and creating another one. We do want to use the daemon if available though.

A solution could be to extend `NodeStrategy`'s API to ensure you're dealing with a direct node in case you have to. Something like a `force_direct()` method (or a context manager?). WDYT?

Comment (dirk-thomas) [reply ros2topic/ros2topic/verb/echo.py:101]: > A solution could be to extend NodeStrategy's API to ensure you're dealing with a direct node in case you have to.

What is the difference to a `DirectNode` then?

File: ros2component/ros2component/api/__init__.py (hunk 1 above)
Comment (dirk-thomas) [reply ros2component/ros2component/api/__init__.py:84]: I don't mind the value. For this PR you can choose whatever.

The problem is that with the current API all waits happen sequentially. E.g. if the service of `N` nodes isn't available the CLI will hang for `N * timeout`. I would prefer that the CLI only blocks for `timeout` seconds (whatever that value is). But that is pretty much our of scope for this PR.

File: ros2topic/ros2topic/verb/echo.py (hunk 2 above)
Comment (dirk-thomas) [reply ros2topic/ros2topic/verb/echo.py:101]: I think this should create direct node, and conditionally if a daemon is running query some information from that to not have to wait for "complete" discovery for the direct node.

Comment (hidmic) [reply ros2topic/ros2topic/verb/echo.py:101]: > What is the difference to a `DirectNode` then?

That it'd use the daemon outside the scope where you actually need it to be direct, e.g.:

```python
with NodeStrategy(args) as node:
    # ...
    with node.force_direct():
        node.subscribe()
```

> I think this should create direct node, and conditionally if a daemon is running query some information from that to not have to wait for "complete" discovery for the direct node.

That sounds a lot like replicating `NodeStrategy`'s implementation N times. I'd rather improve it in a single place.

Comment (hidmic) [reply ros2topic/ros2topic/verb/echo.py:101]: See https://github.com/ros2/ros2cli/pull/499 for one alternative.

Comment (dirk-thomas) [reply ros2topic/ros2topic/verb/echo.py:101]: > See #499 for an alternative.

This looks much better :+1:

**Comment by hidmic** [comment]
CI up to `ros2topic`, `ros2service`, `ros2action`, `ros2node` and `ros2component`:

- Linux [Build Status](https://ci.ros2.org/job/ci_linux/10697/)

File: ros2component/ros2component/api/__init__.py (hunk 1 above)
Comment (dirk-thomas) [reply ros2component/ros2component/api/__init__.py:84]: Please create a follow up ticket for this.

File: ros2component/ros2component/verb/standalone.py [hunk 4]
```diff
@@ -16,8 +16,8 @@
 import signal
 import uuid
 
+from ros2cli.node.direct import add_arguments
 from ros2cli.node.direct import DirectNode
-from ros2cli.node.strategy import add_arguments
```
Comment (dirk-thomas) [inline ros2component/ros2component/verb/standalone.py:20]: This change doesn't seem to be related to the topic of the PR. Please consider separating it.

File: ros2component/ros2component/verb/types.py [hunk 5]
```diff
@@ -25,7 +24,6 @@ class TypesVerb(VerbExtension):
     """Output a list of components registered in the ament index."""
 
     def add_arguments(self, parser, cli_name):
-        add_arguments(parser)
```
Comment (dirk-thomas) [inline ros2component/ros2component/verb/types.py:28]: This change doesn't seem to be related to the topic of the PR. Please consider separating it.

File: ros2service/ros2service/verb/find.py [hunk 6]
```diff
@@ -34,6 +35,7 @@ def add_arguments(self, parser, cli_name):
         parser.add_argument(
             '--include-hidden-services', action='store_true',
             help='Consider hidden services as well')
+        add_strategy_node_arguments(parser)
```
Comment (dirk-thomas) [inline ros2service/ros2service/verb/find.py:38]: This change doesn't seem to be related to the topic of the PR. Please consider separating it.

File: ros2service/ros2service/verb/list.py [hunk 7]
```diff
@@ -22,7 +22,7 @@ class ListVerb(VerbExtension):
     """Output a list of available services."""
 
     def add_arguments(self, parser, cli_name):
-        add_arguments(parser)
+        add_strategy_node_arguments(parser)
```
Comment (dirk-thomas) [inline ros2service/ros2service/verb/list.py:25]: This change doesn't seem to be related to the topic of the PR. Please consider separating it.

File: ros2service/ros2service/verb/type.py [hunk 8]
```diff
@@ -27,6 +28,7 @@ def add_arguments(self, parser, cli_name):
             help="Name of the ROS service to get type (e.g. '/talker/list_parameters')")
         arg.completer = ServiceNameCompleter(
             include_hidden_services_key='include_hidden_services')
+        add_strategy_node_arguments(parser)
```
Comment (dirk-thomas) [inline ros2service/ros2service/verb/type.py:31]: This change doesn't seem to be related to the topic of the PR. Please consider separating it.

File: ros2topic/ros2topic/verb/find.py [hunk 9]
```diff
@@ -33,6 +35,7 @@ def add_arguments(self, parser, cli_name):
         parser.add_argument(
             '--include-hidden-topics', action='store_true',
             help='Consider hidden topics as well')
+        add_strategy_node_arguments(parser)
```
Comment (dirk-thomas) [inline ros2topic/ros2topic/verb/find.py:38]: This change doesn't seem to be related to the topic of the PR. Please consider separating it.

File: ros2topic/ros2topic/verb/list.py [hunk 10]
```diff
@@ -22,7 +23,7 @@ class ListVerb(VerbExtension):
     """Output a list of available topics."""
 
     def add_arguments(self, parser, cli_name):
-        add_arguments(parser)
+        add_strategy_node_arguments(parser)
```
Comment (dirk-thomas) [inline ros2topic/ros2topic/verb/list.py:26]: This change doesn't seem to be related to the topic of the PR. Please consider separating it.

File: ros2component/ros2component/api/__init__.py (hunk 1 above)
Comment (hidmic) [reply ros2component/ros2component/api/__init__.py:84]: Oh, good point. See https://github.com/ros2/ros2cli/issues/507.

File: ros2component/ros2component/verb/standalone.py (hunk 4 above)
Comment (hidmic) [reply ros2component/ros2component/verb/standalone.py]: Unrelated, yes. Reverted in https://github.com/ros2/ros2cli/pull/494/commits/69363670e4c14ede0a1cd48e80cd4b4dce4fed2f. Same for all other cases below.

File: ros2topic/ros2topic/verb/list.py (hunk 10 above)
Comment (jacobperron) [reply ros2topic/ros2topic/verb/list.py:26]: I think this comment still applies.

Comment (hidmic) [reply ros2topic/ros2topic/verb/list.py]: Argh, I thought I had changed them all. See https://github.com/ros2/ros2cli/pull/494/commits/ba6248b11201436e21e9cccee83ac291e29698c2.

File: ros2component/ros2component/api/__init__.py [hunk 11]
```diff
@@ -82,7 +81,10 @@ def get_container_components_info(*, node, remote_container_node_name):
...
+            raise RuntimeError(
+                f"No list_nodes service found for '{remote_container_node_name}' container"
```
Comment (dirk-thomas) [inline ros2component/ros2component/api/__init__.py:86]: The service name should be quoted for better readability.

Same below.

File: ros2topic/ros2topic/verb/list.py [hunk 12]
```diff
@@ -22,7 +22,6 @@ class ListVerb(VerbExtension):
     """Output a list of available topics."""
 
     def add_arguments(self, parser, cli_name):
-        add_arguments(parser)
```
Comment (dirk-thomas) [inline ros2topic/ros2topic/verb/list.py:25]: This change doesn't seem to be related to the topic of the PR. Please consider separating it.

**Comment by dawonn-haval** [comment]
Just wanted to chime in and mention that this patch resolves the cli instability for me across multiple hosts that I've been seeing for a long time. 

I wonder if the same changes need to be made to the  ros2bag package?

File: ros2component/ros2component/api/__init__.py (hunk 11 above)
Comment (hidmic) [reply ros2component/ros2component/api/__init__.py]: See 5a28d9da97d673a795a3e911cd6e843765399065.

File: ros2topic/ros2topic/verb/list.py (hunk 12 above)
Comment (hidmic) [reply ros2topic/ros2topic/verb/list.py]: See 5a28d9da97d673a795a3e911cd6e843765399065.

**Comment by hidmic** [comment]
Running CI once again, got linter test failures due to recent `flake8` updates.

**Comment by clalancette** [comment]
I might also suggest a full CI run on Linux, just to see what the system_tests stuff looks like with this in place.

**Comment by dirk-thomas** [comment]
> I might also suggest a full CI run on Linux, just to see what the system_tests stuff looks like with this in place.

Which part of `ros2cli` do the `system_tests` use?

**Comment by clalancette** [comment]
> Which part of `ros2cli` do the `system_tests` use?

I'm honestly not sure, but this is a big enough change that it makes me wary of the overnights.  I'd prefer to see what it does on a full run, as I'd like not to regress at this point (we are actually getting pretty close to green for Linux CI, for instance).

**Comment by dirk-thomas** [comment]
> I'd prefer to see what it does on a full run,

A more complete run always makes sense. But also only for downstream packages which actually depend on something in the `ros2cli` repo. And in `system_tests` the only package depending on `ros2cli` is [test_security](https://github.com/ros2/system_tests/blob/2f157db5cf47abaf8f3ef0b163c78c57a6dba14f/test_security/package.xml#L29) - so running CI on e.g. `test_communication` would be not helpful and a waste of time / resources.

**Comment by hidmic** [comment]
CI up to and above `ros2cli`, `ros2topic`, `ros2service`, `ros2action`, `ros2node` and `ros2component`:

- Linux [Build Status](https://ci.ros2.org/job/ci_linux/10729)

**Comment by hidmic** [comment]
@clalancette @dirk-thomas all test failures here are unrelated to this patch. I've sent separate PRs to fix flake8 errors and `sros2` test failures. Do you want to move forward or do you want to wait for those to land and make another CI run, this time across platforms?

**Comment by hidmic** [comment]
> It would be good to investigate why the test failure happened in the PR build.

It's a strange one, I cannot reproduce locally.

File: ros2topic/ros2topic/verb/echo.py [hunk 13]
```diff
@@ -105,33 +107,8 @@ def subscriber(
...
-            if n == expanded_name:
-                if len(t) > 1:
-                    raise RuntimeError(
-                        "Cannot echo topic '%s', as it contains more than one type: [%s]" %
-                        (topic_name, ', '.join(t))
-                    )
-                message_type = t[0]
-                break
-        else:
-            raise RuntimeError(
```
Comment (jacobperron) [inline ros2topic/ros2topic/verb/echo.py:128]: I think we need to add back a similar guard to avoid passing `None` to `rclpy.Node.create_subscription`.
This introduced a bug reported in https://github.com/ros2/ros2cli/issues/519

Comment (jacobperron) [reply ros2topic/ros2topic/verb/echo.py:128]: See fix in https://github.com/ros2/ros2cli/pull/520

**Comment by aitazhixin** [comment]
In dashing 0.7.11 version, 'Could not determine the type for the passed topic' also happens.

**Comment by hidmic** [comment]
@aitazhixin I don't think these changes were backported to Dashing. It doesn't seem like it'd hurt either, as it only extends API. @nuclearsandwich thoughts?

**Comment by aitazhixin** [comment]
> @aitazhixin I don't think these changes were backported to Dashing. This patch (and connected ones) only extends and improves API, but I will say that anything bigger than a bugfix is less likely to be considered for a backport. @nuclearsandwich thoughts?

image
so, I need a version which is bigger than 0.9.3?

**Comment by aitazhixin** [comment]
@hidmic so, I need a version which is bigger than 0.9.3?
image
